src/services/transaction_watcher_service.ts

```diff
@@ -0,0 +1,116 @@
+            where: [
+                { status: TransactionStates.Unsubmitted },
+                { status: TransactionStates.Submitted },
+                { status: TransactionStates.Mempool },
+                { status: TransactionStates.Stuck },
+            ],
+        });
+        logger.trace(`found ${transactionsToCheck.length} transactions to check status`);
+        for (const tx of transactionsToCheck) {
+            await this._findTransactionStatusAndUpdateAsync(tx);
+        }
+    }
+    private async _findTransactionStatusAndUpdateAsync(txEntity: TransactionEntity): Promise<TransactionEntity> {
+        // TODO(oskar) - LessThan does not work on dates in TypeORM queries,
+        // ref: https://github.com/typeorm/typeorm/issues/3959
+        const now = new Date();
+        const isExpired = txEntity.expectedAt <= now;
+        try {
+            const txInBlockchain = await this._web3Wrapper.getTransactionByHashAsync(txEntity.hash);
+            if (txInBlockchain !== undefined && txInBlockchain !== null && txInBlockchain.hash !== undefined) {
+                if (txInBlockchain.blockNumber !== null) {
+                    logger.trace({
+                        message: `a transaction with a ${
+                            txEntity.status
+                        } status is already on the blockchain, updating status to confirmed`,
+                        hash: txInBlockchain.hash,
+                    });
+                    txEntity.status = TransactionStates.Confirmed;
+                    txEntity.blockNumber = txInBlockchain.blockNumber;
+                    await this._transactionRepository.save(txEntity);
+                    await this._abortTransactionsWithTheSameNonceAsync(txEntity);
+                    return txEntity;
+                    // Checks if the txn is in the mempool but still has it's status set to Unsubmitted or Submitted
+                } else if (!isExpired && txEntity.status !== TransactionStates.Mempool) {
+                    logger.trace({
+                        message: `a transaction with a ${
+                            txEntity.status
+                        } status is pending, updating status to mempool`,
+                        hash: txInBlockchain.hash,
+                    });
+                    txEntity.status = TransactionStates.Mempool;
+                    return this._transactionRepository.save(txEntity);
+                } else if (isExpired) {
+                    // NOTE(oskar): we currently cancel all transactions that are in the
+                    // "stuck" state. A better solution might be to unstick
+                    // transactions one by one and observing if they unstick the
+                    // subsequent transactions.
+                    txEntity.status = TransactionStates.Stuck;
+                    return this._transactionRepository.save(txEntity);
+                }
+            }
+        } catch (err) {
+            if (err instanceof TypeError) {
```

Comment:
Let's add a `// HACK(oskar): web3Wrapper.getTransactionByHashAsync throws a TypeError if the Ethereum node cannot find the transaction and returns NULL instead of the transaction object. We therefore use this to detect this case until @0x/web3-wrapper is fixed.`

Reply:
Changed the comment from:
```
                // TODO(oskar) `getTransactionByHashAsync` throws a TypeError
                // when tx returned from `eth_getTransactionByHash` is null.
```
To the `HACK` comment.